src/meta/ActiveHostsMan.cpp

```diff
@@ -34,10 +34,20 @@ ActiveHostsMan::ActiveHostsMan(int32_t intervalSeconds, int32_t expiredSeconds,
 bool ActiveHostsMan::updateHostInfo(const HostAddr& hostAddr, const HostInfo& info) {
     std::vector<kvstore::KV> data;
     {
-        folly::RWSpinLock::ReadHolder rh(&lock_);
+        using ReadHolder = folly::RWSpinLock::ReadHolder;
+        using WriteHolder = folly::RWSpinLock::WriteHolder;
+        using UpgradedHolder = folly::RWSpinLock::UpgradedHolder;
+        // Acquire the read lock
+        ReadHolder rh(&lock_);
         auto it = hostsMap_.find(hostAddr);
         if (it == hostsMap_.end()) {
-            folly::RWSpinLock::UpgradedHolder uh(&lock_);
+            // Step to Upgraded phase to stop new readers
+            UpgradedHolder uh(&lock_);
+            // Unlock the read lock
+            rh.reset();
```

Comment:
Is hostsMap_  the  single-writer-many-reader?

Reply:
It is multi-writer, update by heartbeat.

Reply:
No, it's a multi-writer multi-reader case